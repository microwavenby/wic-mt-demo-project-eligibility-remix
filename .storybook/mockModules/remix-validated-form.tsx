import { Form } from "@remix-run/react";
export {
  validationError,
  setFormDefaults,
  createValidator,
  FieldErrors,
  Validator,
} from "../../node_modules/remix-validated-form";
// This is duplicative of the jest helper remixValidatedFormMock,
// but the Jest global missing breaks the component in Storybook

// This ends up being <form> in ./remix-react.tsx
export const ValidatedForm = Form;

type MinimalInputProps = {
  onChange?: (...args: any[]) => void;
  onBlur?: (...args: any[]) => void;
  defaultValue?: any;
  defaultChecked?: boolean;
  name?: string;
  type?: string;
};

export const getInputProps = <T extends MinimalInputProps>(
  props = {} as any
) => {
  return props as T;
};

// TODO: Figure out a way to determine whether this component is being
// rendered in a story with "error" in the name or route, so that
// an example error can be shown in Storybook
export const useField = (args: any) => {
  return { error: undefined, getInputProps };
};
