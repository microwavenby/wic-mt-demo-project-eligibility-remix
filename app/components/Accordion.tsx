import { Trans } from "react-i18next";
import { ReactElement } from "react";
import { Accordion as USWDS_Accordion } from "@trussworks/react-uswds";
import { i18nKey } from "app/types";

export type AccordionProps = {
  bodyKey: i18nKey;
  headerKey: i18nKey;
  id: string;
};

const Accordion = (props: AccordionProps): ReactElement => {
  const { bodyKey, headerKey, id } = props;

  return (
    <USWDS_Accordion
      bordered={true}
      items={[
        {
          title: <Trans i18nKey={headerKey} />,
          content: <Trans i18nKey={bodyKey} />,
          id: id,
          expanded: false,
          headingLevel: "h3",
        },
      ]}
    />
  );
};

export default Accordion;
