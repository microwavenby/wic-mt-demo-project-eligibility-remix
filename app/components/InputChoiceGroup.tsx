import { useTranslation } from "react-i18next";
import { ReactElement } from "react";
import {
  Fieldset,
  Checkbox,
  Radio,
  ErrorMessage,
} from "@trussworks/react-uswds";
import Required from "app/components/Required";
import { useField } from "remix-validated-form";
import { i18nKey } from "app/types";

export type Choice = {
  labelKey: i18nKey;
  name: string;
  value: string;
};

export type InputChoiceGroupProps = {
  choices: Choice[];
  titleKey: i18nKey;
  required?: boolean;
  type: "checkbox" | "radio";
  error?: string;
  helpElement?: ReactElement;
};

const InputChoiceGroup = (props: InputChoiceGroupProps): ReactElement => {
  const { choices, titleKey, required, type, helpElement } = props;
  const { getInputProps, error } = useField(choices[0].name);
  const InputTypeClass = type == "checkbox" ? Checkbox : Radio;
  let { t } = useTranslation();
  return (
    <Fieldset legend={t(titleKey)} legendStyle="srOnly">
      <h2>
        {t(titleKey)}
        {required && <Required />}
      </h2>
      {error && (
        <ErrorMessage id="${titleKey}-error-message">{error}</ErrorMessage>
      )}
      {helpElement}
      {choices.map((choice: Choice) => (
        <InputTypeClass
          value={choice.value}
          tile={true}
          key={`${choice.name}-${choice.value}`}
          {...getInputProps({
            id: `${choice.name}-${choice.value}`,
            label: t(choice.labelKey),
          })}
        />
      ))}
    </Fieldset>
  );
};

export default InputChoiceGroup;
