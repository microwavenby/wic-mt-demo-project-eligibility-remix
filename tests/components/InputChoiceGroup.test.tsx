import { screen } from "@testing-library/react";
import { renderWithRouter } from "tests/helpers/setup";
import { ChangeEvent } from "react";

import InputChoiceGroup, {
  InputChoiceGroupProps,
  Choice,
} from "app/components/InputChoiceGroup";
import Accordion from "app/components/Accordion";
import { useField } from "remix-validated-form";
jest.mock("remix-validated-form");
import { mockUseFieldReturnValue } from "tests/helpers/remixValidatedFormMock";

const mockedUseField = jest.mocked(useField);
mockedUseField.mockReturnValue(mockUseFieldReturnValue);

const choices: Choice[] = ["a", "b", "c"].map((option) => {
  return {
    value: option,
    labelElement: <div>label {option}</div>,
  };
});

const testProps: InputChoiceGroupProps = {
  choices: choices,
  name: "test-choices",
  titleKey: "title",
  required: false,
  type: "checkbox",
  handleChange: (e: ChangeEvent<HTMLInputElement>) => {},
};

it("should match snapshot when it is a set of checkboxes", () => {
  const { container } = renderWithRouter(<InputChoiceGroup {...testProps} />);
  expect(container).toMatchSnapshot();
});

it("should match snapshot when it is a set of radio buttons", () => {
  const { container } = renderWithRouter(
    <InputChoiceGroup {...testProps} type="radio" />
  );
  expect(container).toMatchSnapshot();
});

it("should match display required marker if required is true", () => {
  renderWithRouter(<InputChoiceGroup {...testProps} required={true} />);
  const required = screen.getByText("*");
  expect(required).toBeInTheDocument;
});

it("should display an helpElement if passed as props", () => {
  renderWithRouter(
    <InputChoiceGroup
      {...testProps}
      helpElement={
        <Accordion
          headerKey="accordion header"
          bodyKey="accordion body"
          id="test-accordion"
        />
      }
    />
  );
  const accordionElement = screen.getByText("accordion header");
  expect(accordionElement).toBeInTheDocument;
});
