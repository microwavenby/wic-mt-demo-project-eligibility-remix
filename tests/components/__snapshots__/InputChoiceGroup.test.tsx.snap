// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when it is a set of checkboxes 1`] = `
<div>
  <fieldset
    class="usa-fieldset"
    data-testid="fieldset"
  >
    <legend
      class="usa-sr-only"
    >
      title
    </legend>
    <h2>
      title
    </h2>
    <div
      class="usa-checkbox"
      data-testid="checkbox"
    >
      <input
        class="usa-checkbox__input usa-checkbox__input--tile"
        id="test-choices-a"
        type="checkbox"
        value="a"
      />
      <label
        class="usa-checkbox__label"
        for="test-choices-a"
      >
        <div>
          label 
          a
        </div>
      </label>
    </div>
    <div
      class="usa-checkbox"
      data-testid="checkbox"
    >
      <input
        class="usa-checkbox__input usa-checkbox__input--tile"
        id="test-choices-b"
        type="checkbox"
        value="b"
      />
      <label
        class="usa-checkbox__label"
        for="test-choices-b"
      >
        <div>
          label 
          b
        </div>
      </label>
    </div>
    <div
      class="usa-checkbox"
      data-testid="checkbox"
    >
      <input
        class="usa-checkbox__input usa-checkbox__input--tile"
        id="test-choices-c"
        type="checkbox"
        value="c"
      />
      <label
        class="usa-checkbox__label"
        for="test-choices-c"
      >
        <div>
          label 
          c
        </div>
      </label>
    </div>
  </fieldset>
</div>
`;

exports[`should match snapshot when it is a set of radio buttons 1`] = `
<div>
  <fieldset
    class="usa-fieldset"
    data-testid="fieldset"
  >
    <legend
      class="usa-sr-only"
    >
      title
    </legend>
    <h2>
      title
    </h2>
    <div
      class="usa-radio"
      data-testid="radio"
    >
      <input
        class="usa-radio__input usa-radio__input--tile"
        id="test-choices-a"
        type="radio"
        value="a"
      />
      <label
        class="usa-radio__label"
        for="test-choices-a"
      >
        <div>
          label 
          a
        </div>
      </label>
    </div>
    <div
      class="usa-radio"
      data-testid="radio"
    >
      <input
        class="usa-radio__input usa-radio__input--tile"
        id="test-choices-b"
        type="radio"
        value="b"
      />
      <label
        class="usa-radio__label"
        for="test-choices-b"
      >
        <div>
          label 
          b
        </div>
      </label>
    </div>
    <div
      class="usa-radio"
      data-testid="radio"
    >
      <input
        class="usa-radio__input usa-radio__input--tile"
        id="test-choices-c"
        type="radio"
        value="c"
      />
      <label
        class="usa-radio__label"
        for="test-choices-c"
      >
        <div>
          label 
          c
        </div>
      </label>
    </div>
  </fieldset>
</div>
`;
